feat(server): Flesh out global config service

TBS1996 opened the pull request:
This PR introduces a new global config service to fetch global configs from upstream. The global config is a project-independent config for Relay that may include configuration for Relays or specific to projects. The service is responsible for fetching the config from upstream regularly in order to have an up-to-date version, currently set to the arbitrary number 10 to have a good balance of the latest version without smashing upstream.

Global configs are served from the same endpoint as project configs for practical reasons (e.g. no need to add new routes to proxy allow-lists) with a new parameter, as in the following example. The endpoint version is currently a version above the project configs while we're in development, these two will match once the initiative is closed.

The global config service supports two ways to fetch the global config to the rest of Relay: serving one-off requests and subscribing for continued updates. In this PR, the envelope processor service (subscribed) and the endpoint (one-off requests, to serve the configs downstream) are the only entities using the global config service.

### Example request

```txt
/api/0/relays/projectconfigs/?version=4
{"publicKeys": [1, 2, 3], "global": true}
```

### Future work
This PR also introduces unfinished work that needs further discussion, brought out of the scope of this PR:
- [ ] https://github.com/getsentry/relay/pull/2411
- [ ] https://github.com/getsentry/relay/pull/2445

iker-barriocanal committed:
feat(dynamic-config): Introduce Relay global config

iker-barriocanal committed:
add module and docstrings

iker-barriocanal committed:
feat(services): Create service for global configs

iker-barriocanal committed:
make global config mod public

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
Merge branch 'master' into tor/globalconfig-service

TBS1996 committed:
wip

TBS1996 commented on relay-dynamic-config/src/global.rs:
```diff
@@ -12,11 +12,11 @@ use crate::TaggingRule;
     /// the event's content.
-    #[serde(skip_serializing_if = "Option::is_none")]
-    metric_conditional_tagging: Option<Vec<TaggingRule>>,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
```
I think vectors dont need to be optional wrt version compatibility right? since it'll just use an empty one if its missing

TBS1996 replied:
maybe i need to add a [default] here too

jjbayer replied:
The `#[serde(default)]` on the `struct` should be enough:

> When deserializing, any missing fields should be filled in from the struct's implementation of Default. Only allowed on structs.

https://serde.rs/container-attrs.html

TBS1996 commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    fn path(&self) -> Cow<'static, str> {
+        Cow::Borrowed("/api/0/relays/projectconfigs/?version=4&global_config=true")
```
should match my sentry side change: https://github.com/getsentry/sentry/pull/52421/files#diff-1f1d2ba0fbdd84250e074bd0ec21f8269957b5128716d2e6c3774d705f73d98dR54

jjbayer replied:
Not sure what the current state of the proposal is, but in the last discussion with @olksdr and @iker-barriocanal we decided to put the request for `"global": true` in the request body (not as a query parameter).

olksdr replied:
Yeah, if I remember correctly we did agree to put this key into the body of the request.

TBS1996 commented on relay-server/src/actors/processor.rs:
```diff
@@ -2092,10 +2092,10 @@ impl EnvelopeProcessorService {
 
                     // TODO: Move conditional tagging to generic metrics extraction
-                    let tagging_config = &state.project_state.config.metric_conditional_tagging;
+                    let tagging_config = state.project_state.metric_conditional_tagging();
```
this is where we call our custom method that merges the global and project values. I wish we could make it impossible to get the project values directly, but IIRC, mixed visibility of fields in a struct is against our guidelines

jjbayer replied:
I would make the struct field private nonetheless.

TBS1996 commented on relay-server/src/actors/project.rs:
```diff
@@ -338,6 +344,20 @@ impl ProjectState {
         self.config.features.0.contains(&feature)
     }
+
+    pub fn measurements(&self) -> Option<&MeasurementsConfig> {
```
project override global

olksdr replied:
Please also add docs to newly added functions.

TBS1996 commented on relay-server/src/actors/project.rs:
```diff
@@ -338,6 +344,20 @@ impl ProjectState {
+            .or(self.global_config.measurements.as_ref())
+    }
+
+    pub fn metric_conditional_tagging(&self) -> impl Iterator<Item = &TaggingRule> {
```
please double check if this is the order we want, I vaguely remember that global rules should come first

jjbayer replied:
It's much worse than that: If you look at https://github.com/getsentry/sentry/blob/79bd47dcb793241611584deca4b0f3a9128dc961/src/sentry/relay/config/metric_extraction.py#L150-L175

the list of rules contains:

1. One block of project-specific rules.
2. One block of rules that are either project-specific or a global default.
3. One block of rules that are truly global.

So I would keep the relay-side logic simple and only fall back to the global config when the value is missing in the project config (note that missing is not the same as empty vec -- an empty vec should still mean no tagging rules).

TBS1996 commented on relay-server/src/actors/project_cache.rs:
```diff
@@ -333,14 +344,20 @@ impl ProjectSource {
         let local_source = LocalProjectSourceService::new(config.clone()).start();
-        let upstream_source =
-            UpstreamProjectSourceService::new(config.clone(), upstream_relay).start();
+        let global_config = Arc::new(GlobalConfig::default());
```
this empty global config will hopefully be overridden quickly, but i'd prefer we either don't start any services until we have the first globalconfig fetched from upstream, or we have some kinda static globalconfig that will be used at start-up

olksdr replied:
We should not assume  anything here, and just do not process anything till the global config is in there. 
Instead of default here we might be able to use something like `enum` with different states, one is `init` or `empty` and another with fetched global config

TBS1996 replied:
instead of dealing with an uninitialized variant in the service, wouldn't it be better to wait for the first instance of globalconfig in servicestate::start, if we are not gonna start relay without a fresh globalconfig anyway?

olksdr replied:
We start Relay currently without project configs as well, and also should start the Relay without global config. 
Fetching can take very long time, or upstream currently unavailable and it's perfectly fine.  We just keep buffering for time being and once the configs refreshed  we continue processing.

TBS1996 replied:
ok, i think i'll add it to the following PR, for this one i think its best we just check if the fetching and reading and writing works wrt globalconfigs

TBS1996 commented on relay-server/src/actors/project_cache.rs:
```diff
@@ -374,7 +391,13 @@ impl ProjectSource {
+                    .map(|mut state| {
+                        state.global_config = self.global_config.clone();
+                        state
+                    })
```
this is where we inject the global config into the projectstate when it's coming from redis. we only inject when it comes from redis or upstream, if its' from local then it should already have come from redis or upstream so then the global config should already be present. please correct me if im wrong

TBS1996 commented on relay-server/src/actors/project_upstream.rs:
```diff
@@ -58,7 +59,7 @@ impl UpstreamQuery for GetProjectStates {
 
     fn path(&self) -> Cow<'static, str> {
-        Cow::Borrowed("/api/0/relays/projectconfigs/?version=3")
+        Cow::Borrowed("/api/0/relays/projectconfigs/?version=4&global_config=false")
```
I think i'll revert this, as we have to wait first until the sentry side change is deployed. but this is how i think it will look later.

TBS1996 commented on relay-server/src/actors/project_upstream.rs:
```diff
@@ -361,6 +379,7 @@ impl UpstreamProjectSourceService {
                             counter(RelayCounters::ProjectUpstreamCompleted) += 1,
                             result = result,
                         );
+                        state.global_config = self.global_config.clone();
```
this is where the global config is injected when it's coming from the upstream.

TBS1996 commented on relay-server/src/metrics_extraction/conditional_tagging.rs:
```diff
@@ -2,7 +2,11 @@ use relay_dynamic_config::TaggingRule;
-pub fn run_conditional_tagging(event: &Event, config: &[TaggingRule], metrics: &mut [Metric]) {
+pub fn run_conditional_tagging<'a>(
+    event: &Event,
+    config: impl Iterator<Item = &'a TaggingRule>,
```
changed to an iterator because I wanted to merge the global and project tagging rules without any new allocations, lmk if this is overkill

TBS1996 commented on relay-server/src/actors/project_cache.rs:
```diff
@@ -790,6 +818,11 @@ impl ProjectCacheBroker {
     }
 
+    fn handle_update_global_config(&mut self, global_config: Arc<GlobalConfig>) {
+        self.source.upstream_source.send(global_config.clone());
```
ideally I'd send the global config directly from the `GlobalConfigService`, but the `UpstreamProjectSource` service is created within the `ProjectSource` service, and not in `ServiceState::Start` so I don't have access to the Addr when creating the `GlobalConfigService`. meaning I cannot send it directly without refactoring a bunch.

TBS1996 committed:
wip

jjbayer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,105 @@
+/// Service implementing the [`GlobalConfig`] interface.
+///
+/// The service is responsible to fetch the global config appropriately and
+/// forward it to the services that require it.
```
```suggestion
/// The service is responsible for fetching the global config and
/// forwarding it to the services that require it.
```

jjbayer commented on relay-server/src/actors/project.rs:
```diff
@@ -79,6 +82,8 @@ pub struct ProjectState {
     /// The project's current config.
     #[serde(default)]
     pub config: ProjectConfig,
+    #[serde(default)]
```
This definitely needs a doc comment, because it is not obvious to an uninformed reader why the project state contains a reference to global config.

jjbayer commented on relay-server/src/actors/project_upstream.rs:
```diff
@@ -421,15 +440,27 @@ impl UpstreamProjectSourceService {
+            UpstreamProjectSource::FetchProjectState(fetch_project_state, sender) => {
+                self.handle_project_source(fetch_project_state, sender)
+            }
+            UpstreamProjectSource::FetchGlobalConfig(global_config) => {
```
nit: I would call this `UpdateGlobalConfig`. The handler of this message does not fetch anything.

jjbayer commented on relay-server/src/actors/project_upstream.rs:
```diff
@@ -421,15 +440,27 @@ impl UpstreamProjectSourceService {
+                self.handle_project_source(fetch_project_state, sender)
+            }
+            UpstreamProjectSource::FetchGlobalConfig(global_config) => {
+                self.global_config = global_config;
```
I think we need to reconsider allowing different global configs to exist at the same time. Project states may live up to one hour and that's a long time for the global config to go out of date, which will lead to behavior that's hard to explain.

I think we should make this an `Arc<Mutex>` or better yet, publish it not into the project cache but into the service that actually needs it (the `EnvelopeProcessor`). We can still have a somewhat unified interface that is created inside the `EnvelopeProcessor` on the fly.

jjbayer reviewed (commented):
Good general direction, I left some comments.

olksdr commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,105 @@
+
+// No messages are accepted for now.
+#[derive(Debug)]
+pub enum GlobalConfigMessage {}
```
This is basically the name of the `Interface` - this is not a message per se. 
```suggestion
pub enum GlobalConfig {}
```

olksdr commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,105 @@
+}
+
+impl Service for GlobalConfigService {
+    type Interface = GlobalConfigMessage;
```
Also this currently does not support any messages, you could just use `()` for the `Interface`

```suggestion
    type Interface = ();
```

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 commented on relay-dynamic-config/src/project.rs:
```diff
@@ -107,6 +108,32 @@ impl Default for ProjectConfig {
+            .iter()
+            .chain(global_config.metric_conditional_tagging.iter())
+    }
+}
```
regarding the tagging rules that could be either global or dynamic, should we perhaps iterate on those and put them all in dynamic first? @jjbayer

TBS1996 commented on relay-general/src/store/normalize/span/tag_extraction.rs:
```diff
@@ -125,9 +125,9 @@ fn extract_shared_tags(event: &Event) -> BTreeMap<SpanTagKey, String> {
-        if let Some(transaction_method) = transaction_method_from_request
-            .or(http_method_from_transaction_name(transaction_name).map(|m| m.to_uppercase()))
-        {
+        if let Some(transaction_method) = transaction_method_from_request.or_else(|| {
```
this suddenly made me `make lint-rust` fail, even though it got successfully merged in a previous PR. what's up with that?

jjbayer replied:
Not sure, but the change makes sense.

TBS1996 commented on relay-server/src/actors/processor.rs:
```diff
@@ -3340,11 +3381,9 @@ mod tests {
-        };
+        let mut config = ProjectConfig::default();
+        config.datascrubbing_settings = datascrubbing_settings;
+        config.pii_config = Some(pii_config);
```
since there are now private fields in `ProjectConfig`, had to change it like this

TBS1996 committed:
wip

TBS1996 commented on relay-server/src/actors/processor.rs:
```diff
@@ -1351,6 +1369,14 @@ impl EnvelopeProcessorService {
+                // some kinda metric
+                Arc::new(GlobalConfig::default())
+            }
+        };
```
so heres the main change from previous designs, now the `GlobalConfigService` sends to `EnvelopeProcessor` which will put an arc clone into the projectstate. this ensures that it's the same copy of the GlobalConfig that is used in all stages of the envelope processing, since it's an Arc it's also "free" to copy it whenever the globalconfig haven't changed. The RwLock is optimized for reading rather than writing, which is great for us since we do a lot more reading. 

Since the arc is immutable we can also do things like creating iterators over both global and local configs without any new allocations, which is great since the global config is fairly large.  

 we also don't have the problem about an old globalconfig hanging around for  along time like in previous designs, as it gets refreshed on every new envelope

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 commented on relay-dynamic-config/src/project.rs:
```diff
@@ -107,6 +108,32 @@ impl Default for ProjectConfig {
+            .iter()
+            .chain(global_config.metric_conditional_tagging.iter())
+    }
+}
```
since we want to first check if the fetching/reading/writing logic works, we will for this PR not use this merging logic, and rather do it in a following PR

jjbayer replied:
Let's remove it then.

```suggestion
```

TBS1996 commented on relay-server/src/statsd.rs:
```diff
@@ -536,6 +536,18 @@ pub enum RelayCounters {
+    /// Successfully read from the global config.
+    GlobalConfigReadSuccess,
+    /// Successfully read from the global config.
+    GlobalConfigReadFailed,
```
is this a bit overkill just for temporarily checking if it works in prod?

TBS1996 committed:
wip

TBS1996 committed:
Merge branch 'master' into tor/globalconfig-service

TBS1996 commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,105 @@
+    // This struct is only used for getting the global configs, meaning the presence of the key
+    // is enough, and no value needs to be included.
+    global_config: (),
+}
```
is my logic here correct? if we only send this request for global_config, then it is redundant to have something like global_config = true, iiuc sentry will receive the serialized version of this, which will probably have "global_config": null, as the request body, and we only check if the global_config key is present in that case

TBS1996 committed:
wip

TBS1996 committed:
Merge branch 'tor/globalconfig-service' of https://github.com/getsentry/relay into tor/globalconfig-service

TBS1996 committed:
wip

TBS1996 commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    fn path(&self) -> Cow<'static, str> {
+        Cow::Borrowed("/api/0/relays/projectconfigs/")
```
i removed the `version=` thing here, not sure if we should keep it or not, atm it doesn't matter, but in the future we will have different versions of globalconfig. I suppose either we could start from beginning with globalconfig version=1, or if we wanna tightly couple it with the project config version then we will start at 4

TBS1996 committed:
Merge branch 'master' into tor/globalconfig-service

jjbayer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,108 @@
+                    envelope_processor.send::<GlobalConfig>(response.global);
+                }
+                Err(e) => {
+                    relay_log::error!("failed to send request: {}", e);
```
```suggestion
                    relay_log::error!("failed to send global config request: {}", e);
```

jjbayer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,108 @@
+                    relay_log::error!("failed to send request: {}", e);
+                }
+                Ok(Err(e)) => {
+                    relay_log::error!("failed to fetch request: {}", e);
```
```suggestion
                    relay_log::error!("failed to fetch global config request: {}", e);
```

jjbayer commented on relay-server/src/actors/processor.rs:
```diff
@@ -1351,6 +1370,17 @@ impl EnvelopeProcessorService {
+                config.clone()
+            }
+            Err(_) => {
+                metric!(counter(RelayCounters::GlobalConfigReadFailed) += 1);
```
Let's make this one metric with a tag `success = true|false`.

TBS1996 replied:
iiuc, ill have only a `GlobalConfigRead` variant and hten annotate with either success or not, instead of having one variant per outcome?

jjbayer commented on relay-server/src/actors/processor.rs:
```diff
@@ -541,6 +557,7 @@ impl EnvelopeProcessorService {
         project_cache: Addr<ProjectCache>,
         upstream_relay: Addr<UpstreamRelay>,
     ) -> Self {
+        let global_config = RwLock::new(Arc::new(GlobalConfig::default()));
```
For now, we can treat the `GlobalConfig` as optional, but I agree with @olksdr's earlier comments here:
```suggestion
        let global_config = RwLock::new(Arc::new(GlobalConfig::default()));
```

https://github.com/getsentry/relay/pull/2320#discussion_r1267751746

```suggestion
        // TODO: Replace with state enum (`init`, `ready`) and do not process anything while global_config is undefined.
        let global_config = RwLock::new(Arc::new(GlobalConfig::default()));
```

TBS1996 replied:
yep! coming in next PR

jjbayer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,108 @@
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct GetGlobalConfigResponse {
+    global: GlobalConfig,
```
Let's make this an `Arc` from the start. This will allow to send the same config to multiple services without deep copying, and makes later `Arc`ing in the `EnvelopeProcessor` unnecessary.

```suggestion
    global: Arc<GlobalConfig>,
```

TBS1996 replied:
will it implicitly turn it into an Arc here?

```Rust
            match upstream_relay.send(SendQuery(query)).await {
                Ok(Ok(response)) => {
                    envelope_processor.send::<Arc<GlobalConfig>>(response.global);
                }
```

jjbayer reviewed (approved):
No blockers, but please see comments before merging.

TBS1996 committed:
wip

TBS1996 committed:
Merge branch 'tor/globalconfig-service' of https://github.com/getsentry/relay into tor/globalconfig-service

TBS1996 committed:
wip

jjbayer reviewed (changes requested):
Almost forgot: Because we're operating a chain of Relays, you also need to add the ability to serve the global config to downstream Relays here: https://github.com/getsentry/relay/blob/fb9ce18c2b31a90967dfbd7f3970928fff17153e/relay-server/src/endpoints/project_configs.rs.

TBS1996 committed:
Merge branch 'master' into tor/globalconfig-service

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
Merge branch 'master' into tor/globalconfig-service

TBS1996 committed:
wip

TBS1996 committed:
Merge branch 'tor/globalconfig-service' of https://github.com/getsentry/relay into tor/globalconfig-service

TBS1996 committed:
wip

TBS1996 committed:
wip

jjbayer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,111 @@
+#[serde(rename_all = "camelCase")]
+pub struct GetGlobalConfig {
+    global_config: bool,
+}
```
As discussed, this struct needs a dummy `project_keys` field because otherwise upstream Relays won't be able to deserialize it into https://github.com/getsentry/relay/blob/2110e578ac46153d19122a9aa525e1f1c976b5b3/relay-server/src/endpoints/project_configs.rs#L87-L88

Instead of defining a new struct here we can use https://github.com/getsentry/relay/blob/2110e578ac46153d19122a9aa525e1f1c976b5b3/relay-server/src/actors/project_upstream.rs#L34-L38

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,101 @@
+}
+
+/// Service interface for the [`GetGlobalConfig`] message.
+pub struct GlobalConfigResponse(Sender<Arc<GlobalConfig>>);
```
According to conventions this would just be named `GlobalConfig` I believe, like how the projectcache interface is just named `ProjectCache`, but `GlobalConfig` is already the name of the global config itself

iker-barriocanal replied:
If the issue is the name collision, you can name this `GlobalConfig` and use the qualified name for the other one (e.g. `relay_dynamic_config::GlobalConfig`). I don't have strong opinions, so your call on what you like most 😄

TBS1996 replied:
hmm if that's the alternative then I think i'll keep it like it is now 🤔

olksdr replied:
Let's follow the pattern we have in other services and give the interface the proper name. Right now it sounds like this is a message name and not the interface name. 

Maybe something like following would do:

```suggestion
pub struct GlobalConfiguration(Sender<Arc<GlobalConfig>>);
```

And to keep it consistent, you could use the full `GlobalConfiguration` prefix through out the entire file (e.g. `GlobalConfigurationService`  and so on)

TBS1996 replied:
good idea 👍

TBS1996 commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,101 @@
+            full_config: false,
+            no_cache: false,
+            global_config: true,
+        };
```
now we are using the same query both for project cache and global config to the sentry endpoint

TBS1996 commented on relay-server/src/actors/processor.rs:
```diff
@@ -290,6 +292,10 @@ struct ProcessEnvelopeState {
 
+    /// The configuration options for projects which apply to all DSNs
+    /// TODO(tor) implement global config logic
+    _global_config: Arc<GlobalConfig>,
```
Actually using the config will be done in #2364

iker-barriocanal commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,101 @@
+            global_config: true,
+        };
+
+        match upstream_relay.send(SendQuery(query)).await {
```
Does this `await` block the thread until a response is received? If yes, what do you think about spawning another task to do this fetching?

TBS1996 replied:
spawning another task was the previous version, the problem is that we wanna store the most recent globalconfig in the service, and inside of a new task we don't have access to `self`  
 
about if it's blocking or not, im not that familiar with async, i thought it didn't, but even if it does, there is should be a timeout on the fetching requests so it wouldn't block forever

jjbayer replied:
+1 on spawning, inside the `spawn_handler` loop. See for example https://github.com/getsentry/relay/blob/f9421996eb64837ae9dd6816cfe30e6ba4b13ed4/relay-server/src/actors/health_check.rs#L129-L132

TBS1996 replied:
but your example doesn't have mutable access within the spawn? the only way i can think of avoiding this atm is somehow sending it back from `EnvelopeProcessor` which would be very strange

jjbayer replied:
Right, probably a bad example then, sorry. You need to bring the result of `upstream_relay.send(..).await` back into  the spawn handler `loop` where you have mutable access to `self`. Look at how `ProjectCacheBroker` does it: It creates an internal channel, spawns a task that publishes the result of the upstream request into the channel, and reads from the channel in the central loop:

https://github.com/getsentry/relay/blob/71b31159898b69f7a9906dce4acc51d7491a193a/relay-server/src/actors/project_cache.rs#L593-L625

https://github.com/getsentry/relay/blob/71b31159898b69f7a9906dce4acc51d7491a193a/relay-server/src/actors/project_cache.rs#L903

TBS1996 replied:
right, that makes sense! I made it that way now, thx!

iker-barriocanal commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,101 @@
+                        if let Some(sender) = message {
+                            sender.0.send(self.global_config.clone());
+                        }
+                    },
```
What do you think about typing this message and checking it explicitly, instead of always returning the config? Would it be easier to maintain and extend?

TBS1996 replied:
I assumed it was trivial to turn it into an enum when we get multiple  types of messages, are my assumptions wrong ?

iker-barriocanal commented on relay-server/src/service.rs:
```diff
@@ -279,6 +285,11 @@ impl ServiceState {
         &self.inner.registry.processor
     }
 
+    /// Returns the address of the [`GlobalConfigService`] service.
```
Nice!

iker-barriocanal commented on relay-server/src/endpoints/project_configs.rs:
```diff
@@ -84,12 +87,14 @@ struct GetProjectStatesResponseWrapper {
     #[serde(default)]
     no_cache: bool,
+    #[serde(default)]
+    global_config: bool,
```
Based on the `rename_all` in the struct, will this get renamed to `globalConfig` (instead of `global`)? Should we add a rename for it?

TBS1996 replied:
on sentry, the request is `globalConfig` while the response is `global`, which is pretty dumb actually, it wasn't a councious decision. Perhaps we should change that.

TBS1996 replied:
now it's changed!

iker-barriocanal commented on relay-server/src/actors/project_upstream.rs:
```diff
@@ -32,9 +32,10 @@ use crate::utils::{RetryBackoff, SleepHandle};
+    pub public_keys: Vec<ProjectKey>,
+    pub full_config: bool,
+    pub no_cache: bool,
+    pub global_config: bool,
```
IIUC, the project upstream always sends `global_config=false`. Q: do we need to type it explicitly, or can we ignore it?

TBS1996 replied:
it should send =true when we request a global config from GlobalConfigService, maybe i am misunderstanding something 🤔

iker-barriocanal commented on relay-server/src/actors/project_cache.rs:
```diff
@@ -693,6 +693,11 @@ impl ProjectCacheBroker {
+            // processing an envelope. This approach simplifies the processing
+            // of the envelope by abstracting the existence of the global
+            // config.
+
```
I believe this comment is no longer valid and we can get rid of it?

iker-barriocanal commented on relay-server/src/endpoints/project_configs.rs:
```diff
@@ -76,6 +77,8 @@ struct GetProjectStatesResponseWrapper {
     #[serde(skip_serializing_if = "Vec::is_empty")]
     pending: Vec<ProjectKey>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    global: Option<GlobalConfig>,
```
Should we update the docstrings of this function?

TBS1996 replied:
good call!

iker-barriocanal commented on relay-server/src/endpoints/project_configs.rs:
```diff
@@ -122,6 +128,15 @@ async fn inner(
+            .map(|gc| (*gc).clone())
+    } else {
+        None
+    };
```
Does `then` make this cleaner? Something like this (haven't tried it):
```suggestion
    let global = inner.global_config.then(|| {
        global_config_service
            .send(GetGlobalConfig)
            .await
            .ok()
            .map(|gc| (*gc).clone())
    });
```

TBS1996 replied:
i tried a while before i gave up on that approach haha. the main issue is the async within the closure, maybe it's possible but I didn't succeed

iker-barriocanal commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,101 @@
+        let upstream_relay: Addr<UpstreamRelay> = self.upstream.clone();
+        let project_cache = self.project_cache.clone();
+
+        let query = GetProjectStates {
```
Does reusing this object force us to use the same versioning for both the global config and project configs? I think that during development we should use a different version, to make changes and use the global config without changing the project config. What do you think?

TBS1996 replied:
i think we are not gonna forcefully associate a certain globalconfig version with a projectconfig version, we can start requesting specific global_config versions when we make an update to it

jjbayer replied:
Whatever you agree upon, please make sure the solution is synced with https://github.com/getsentry/sentry/pull/53773 (`GetProjectStates` [hard codes](https://github.com/getsentry/relay/blob/f9421996eb64837ae9dd6816cfe30e6ba4b13ed4/relay-server/src/actors/project_upstream.rs#L60-L62) `version=3`).

iker-barriocanal replied:
@TBS1996 Could we use version=4 here?

iker-barriocanal commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,101 @@
+                Some(global_config) => {
+                    let global_config = Arc::new(global_config);
+                    self.global_config = global_config.clone();
+                    project_cache.send::<Arc<GlobalConfig>>(global_config);
```
Q: what does the project cache need the global config for? Isn't the global config service sending the global config to the processor service directly?

TBS1996 replied:
it's there because of the downstream relays, so when a request is sent to this relay, it sends the `GetGlobalConfig` message and returns the most recent global config

olksdr replied:
Same questions as from @iker-barriocanal .
I do not see that `project_cache` even getting the message you send here. How it's being used?
Can actually this service respond to the requests from the upstream relays?

TBS1996 replied:
`project_cache` isn't supposed to get it, I see that i had misnamed the `EnvelopeProcessor`, fixed it now. 

it should be requested from downstream here: https://github.com/getsentry/relay/pull/2320/files#diff-218147664f1c1560b5c41683741a3a186caa23b54ddc86767b21b5fdcbf226bcR134

and sent from here: https://github.com/getsentry/relay/pull/2320/files#diff-9e66add7db47c3f18146ebef0427032761cc7c185870ff3e603e73441e585b2fR92

iker-barriocanal commented on relay-server/src/actors/processor.rs:
```diff
@@ -1360,6 +1383,7 @@ impl EnvelopeProcessorService {
             extracted_metrics: Default::default(),
             project_state,
             sampling_project_state,
+            _global_config: self.global_config.clone(),
```
nit (personal preference): does the compiler scream if we use `Arc::clone` instead of `.clone`? I think it's easier to understand that way. Your call on what your preference is (and what the compiler allows hehe)

TBS1996 replied:
i didn't think about doing it that way, i agree actually, since i prefer `.clone()` to signal something potentially expensive, thanks for the tip! 😄

TBS1996 committed:
wip

olksdr commented on CHANGELOG.md:
```diff
@@ -26,6 +26,7 @@
 
 **Internal**:
 
+- Add new service for fetching global configs. ([#2320](https://github.com/getsentry/relay/pull/2320))
```
This must be in `Unreleased` section

olksdr commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,101 @@
+                    self.global_config = global_config.clone();
+                    project_cache.send::<Arc<GlobalConfig>>(global_config);
+                }
+                None => relay_log::error!("Global config request returned a None value"),
```
Please, give the better error here, why exactly this is `None`. Since current error gives no actual explanation what's going on

TBS1996 replied:
well, we wouldn't really know what happened if it didn't include a global config, is this more clear: `"Upstream response didn't include a global config"`

olksdr commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,101 @@
+}
+
+/// Service interface for the [`GetGlobalConfig`] message.
+pub struct GlobalConfigResponse(Sender<Arc<GlobalConfig>>);
```
Also, please follow the convention and defined the proper messages, which will be more explicit and gives better overview of what operations is it. 
You already have the message defined below, you can just make sure to create an enum and put it inside of the interface.

olksdr commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,101 @@
+    }
+}
+
+impl Interface for GlobalConfigResponse {}
```
Please, move this to below the Interface itself.

olksdr commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,101 @@
+                    _ = ticker.tick() => self.update_global_config().await,
+                    message = rx.recv() => {
+                        if let Some(sender) = message {
+                            sender.0.send(self.global_config.clone());
```
It would be great to follow the convention and add `handle_message` function to the `impl` of the service and do all the messages handling in there. Also explicit type of the message will give better understanding of the code

olksdr commented on relay-server/src/actors/processor.rs:
```diff
@@ -541,6 +562,7 @@ impl EnvelopeProcessorService {
         project_cache: Addr<ProjectCache>,
         upstream_relay: Addr<UpstreamRelay>,
     ) -> Self {
+        // TODO(tor): Replace with state enum (`init`, `ready`) and do not process anything while global_config is undefined.
```
Are you going to add this into this PR? Or in followup?

TBS1996 replied:
followup

jan-auer replied:
Not needed, just make a call on what behavior you want:
 1. (preferred) Delay accepting process messages until the global config is retrieved. 
 2. (dangerous) Process with default config until retrieved.

TBS1996 replied:
the way to go I suppose, would be for the global config service to send an `Arc<Option<GlobalConfig>>`, which only would get the `Some` variant when it has a received from the upstream?

olksdr commented on relay-server/src/actors/processor.rs:
```diff
@@ -2695,29 +2724,38 @@ impl EnvelopeProcessorService {
                 self.handle_rate_limit_flush_buckets(message);
             }
+            // This message is handled in the spawn_handler loop.
+            EnvelopeProcessor::UpdateGlobalConfig(_) => {}
```
Please, add the error here. Just to avoid the problems if someone refactors the code and forgets about this piece, so we at least will catch the error and will know what happens.

olksdr commented on relay-server/src/endpoints/project_configs.rs:
```diff
@@ -70,12 +71,17 @@ impl ProjectStateWrapper {
 /// Version 3 also adds a list of projects whose response is pending.  A [`ProjectKey`] should never
 /// be in both collections.  This list is always empty before V3.
+///
+/// The Response may also have a [`GlobalConfig`] which should be returned if the `global_config`
```
```suggestion
/// The response may also have a [`GlobalConfig`] which should be returned if the `global_config`
```

olksdr commented on relay-server/src/endpoints/project_configs.rs:
```diff
@@ -70,12 +71,17 @@ impl ProjectStateWrapper {
     #[serde(skip_serializing_if = "Vec::is_empty")]
     pending: Vec<ProjectKey>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    global: Option<GlobalConfig>,
```
Can we stay consistent, please? Either call everything `gloabl_config` or `global`. 
```suggestion
    global_config: Option<GlobalConfig>,
```

TBS1996 replied:
ok, I'll change it, added you as reviewer for the sentry side: https://github.com/getsentry/sentry/pull/53918

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip [no ci]

TBS1996 committed:
wip [no ci]

TBS1996 committed:
wip [no ci]

TBS1996 committed:
wip

TBS1996 committed:
wip

jjbayer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,107 @@
+
+impl Interface for GlobalConfiguration {}
+
+pub struct GetGlobalConfig;
```
docs please

jjbayer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,107 @@
+}
+
+impl GlobalConfigurationService {
+    pub fn new(project_cache: Addr<EnvelopeProcessor>, upstream: Addr<UpstreamRelay>) -> Self {
```
docs please

jjbayer replied:
```suggestion
    pub fn new(envelope_procesor: Addr<EnvelopeProcessor>, upstream: Addr<UpstreamRelay>) -> Self {
```

jjbayer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,107 @@
+    fn handle_message(&self, message: GlobalConfiguration) {
+        match message {
+            GlobalConfiguration::GetGlobalConfig(global_config) => {
+                global_config.send(self.global_config.clone());
```
```suggestion
            GlobalConfiguration::GetGlobalConfig(sender) => {
                sender.send(self.global_config.clone());
```

jjbayer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,107 @@
+    /// Forwards the given global config to the services that require it.
+    async fn update_global_config(&mut self) {
+        let upstream_relay: Addr<UpstreamRelay> = self.upstream.clone();
+        let project_cache = self.envelope_processor.clone();
```
```suggestion
        let envelope_processor = self.envelope_processor.clone();
```

jjbayer commented on relay-server/src/actors/project_upstream.rs:
```diff
@@ -48,6 +49,8 @@ pub struct GetProjectStatesResponse {
     #[serde(default)]
     pub pending: Vec<ProjectKey>,
+    #[serde(default)]
+    pub global_config: Option<GlobalConfig>,
```
Let's document these fields, now that they are pub. You can also make them `pub(crate)`.

jjbayer commented on tests/integration/fixtures/mini_sentry.py:
```diff
@@ -339,8 +339,7 @@ def get_project_config():
             abort(500, "unsupported version")
-
-        return jsonify(configs=configs, pending=pending)
+        return jsonify({"configs": configs, "pending": pending, "global": {}})
```
```suggestion
        return jsonify({"configs": configs, "pending": pending, "global_config": {}})
```

TBS1996 committed:
wip

TBS1996 committed:
wip

jjbayer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -26,6 +26,7 @@ pub enum GlobalConfiguration {
 
 impl Interface for GlobalConfiguration {}
 
+/// The message for requesting the most recent global config from [`GlobalConfigurationService`]
```
Other reviewers do this to me all the time, now it's my turn 😉 

```suggestion
/// The message for requesting the most recent global config from [`GlobalConfigurationService`].
```

jjbayer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -37,27 +38,28 @@ impl FromMessage<GetGlobalConfig> for GlobalConfiguration {
 
 impl GlobalConfigurationService {
-    pub fn new(project_cache: Addr<EnvelopeProcessor>, upstream: Addr<UpstreamRelay>) -> Self {
+    /// Creates a new [`GlobalConfigurationService`]
```
```suggestion
    /// Creates a new [`GlobalConfigurationService`].
```

jjbayer commented on relay-server/src/actors/project_upstream.rs:
```diff
@@ -45,12 +45,15 @@ pub struct GetProjectStates {
     #[serde(default)]
-    pub configs: HashMap<ProjectKey, ErrorBoundary<Option<ProjectState>>>,
+    pub(crate) configs: HashMap<ProjectKey, ErrorBoundary<Option<ProjectState>>>,
+    /// The [`ProjectKey`]'s that couldn't be immediatly retrieved from the upstream.
```
```suggestion
    /// The [`ProjectKey`]'s that couldn't be immediately retrieved from the upstream.
```

TBS1996 committed:
Merge branch 'master' into tor/globalconfig-service

TBS1996 committed:
Apply suggestions from code review

Co-authored-by: Joris Bayer <joris.bayer@sentry.io>

TBS1996 committed:
wip

TBS1996 committed:
wip

olksdr commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,121 @@
+/// Service interface for the [`GetGlobalConfig`] message.
+pub enum GlobalConfiguration {
+    GetGlobalConfig(Sender<Arc<GlobalConfig>>),
+}
```
It's a good practice we have to add more documentation to the interface describing what it does and how to use it.
This can include the workflow and the messages it receives or sends to update its state.

olksdr commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,121 @@
+                        };
+                        envelope_processor.send::<Arc<GlobalConfig>>(global_config);
+                    }
+                    None => relay_log::error!("Upstream response didn't include a global config"),
```
I would use something like following:
```suggestion
                    None => relay_log::error!("Global config is missing in upstream response"),
```

just to avoid the `didn't` with single quote in there.

olksdr commented on relay-server/src/actors/processor.rs:
```diff
@@ -290,6 +292,10 @@ struct ProcessEnvelopeState {
     /// This is the config used for trace-based dynamic sampling.
     sampling_project_state: Option<Arc<ProjectState>>,
 
+    /// The configuration options for projects which apply to all DSNs
```
```suggestion
    /// The configuration options for projects which apply to all DSNs.
```

olksdr commented on relay-server/src/actors/project_upstream.rs:
```diff
@@ -44,10 +45,15 @@ pub struct GetProjectStates {
+    /// The global config fetched from the upstream.
     #[serde(default)]
-    pub pending: Vec<ProjectKey>,
+    pub(crate) global: Option<GlobalConfig>,
```
Please, keep the naming consistent. Either `global` everywhere or `global_config`. 

```suggestion
    pub(crate) global_config: Option<GlobalConfig>,
```

TBS1996 committed:
Apply suggestions from code review

Co-authored-by: Oleksandr <1931331+olksdr@users.noreply.github.com>

iker-barriocanal reviewed (commented):
The PR overall looks good to me. I think we can benefit from two changes before landing this PR:
- Using a different version mechanism for global configs than project configs (for `version=4`).
- Typing `global` in the protocol.

TBS1996 committed:
wip

jjbayer reviewed (changes requested):
As discussed:

* Bump endpoint version to 4 (requires changes in sentry and relay)
* Align field names in request and response bodies (requires changes in sentry and relay)
* Respond with a 5xx response when fetching from the service fails
* Make sure that unauthenticated relays do not attempt to fetch global config

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
Merge branch 'master' into tor/globalconfig-service

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
Merge branch 'master' into tor/globalconfig-service

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

iker-barriocanal commented on .foorelay/fooconfig.yml:
What are these files for?

iker-barriocanal commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,138 @@
+        envelope_processor: Addr<EnvelopeProcessor>,
+        upstream: Addr<UpstreamRelay>,
+    ) -> Self {
+        let global_config = Arc::new(GlobalConfig::default());
```
How does the service behave when a global config is yet not received (i.e. we still have the default global config here), and Relay gets a request from downstream to fetch the global config?

iker-barriocanal reviewed (changes requested):
Main approach looks good!

TBS1996 committed:
wip

iker-barriocanal committed:
feat: global config subscription mechanism

using a watch channel to make the solution generic

iker-barriocanal commented on relay-server/src/endpoints/project_configs.rs:
```diff
@@ -150,7 +164,11 @@ async fn inner(
+    }))
 }
 
 /// Returns `true` if the `?version` query parameter is compatible with this implementation.
```
Let's make v4 compatible with this Relay.

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
Merge branch 'master' into tor/globalconfig-service

TBS1996 committed:
Merge branch 'tor/globalconfig-service' of https://github.com/getsentry/relay into tor/globalconfig-service

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
revert relay version

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 commented on relay-dynamic-config/src/global.rs:
```diff
@@ -7,16 +7,25 @@ use crate::TaggingRule;
+            measurements: Some(MeasurementsConfig::default()),
+            metric_conditional_tagging: Default::default(),
+        }
+    }
```
added this just for the integration test, otherwise just an empty dict is returned since we don't have a proper default yet, which makes it hard to know if the integration test is succesful

TBS1996 commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,143 @@
+                            relay_log::error!("failed to update global config watch: {}", e);
+                        };
+                    },
+                    _ = ticker.tick(), if self.enabled =>  {
```
its not enough to just check the upstream authentication, because if it's not in managed mode, it will say it's authenticated even though it's not, so that's why i have an additional `enabled` flag which for now just checks if relay credentials are present. reading through the code, it seems credentials do not change during normal operation.

TBS1996 commented on relay-server/src/actors/processor.rs:
```diff
@@ -563,12 +574,14 @@ impl EnvelopeProcessorService {
         };
 
         Self {
+            global_config: Arc::new(GlobalConfig::default()),
```
will be a proper default in next pr, we should discuss a way to ensure the default values in sentry and the ones in relay are synchronized though, when it comes to the values that are unchanging

TBS1996 commented on relay-server/src/actors/processor.rs:
```diff
@@ -2750,22 +2764,39 @@ impl EnvelopeProcessorService {
+                .await
+                .expect(
+                    "envelope processorservice failed to subscribe to global configurationservice",
+                );
```
we should discuss the correct error handling here. But at least this one will only fail when relay is starting

TBS1996 commented on relay-server/src/actors/processor.rs:
```diff
@@ -2750,22 +2764,39 @@ impl EnvelopeProcessorService {
+                   _ = config_rx.changed() => self.global_config = config_rx.borrow().clone(),
+
+                    (Some(message), Ok(permit)) =
+                        async {tokio::join!(rx.recv(), semaphore.clone().acquire_owned())} => {
```
this was the best way that me and iker found to put a join in a select, the problem with join in select is that select expects a future, but join evaluates to a tuple of the result of the futures

TBS1996 commented on tests/integration/fixtures/mini_sentry.py:
```diff
@@ -308,17 +308,25 @@ def get_project_config():
+            if version == "4":
+                if flask_request.json.get("global") or True:
+                    global_key["measurements"] = {"maxCustomMeasurements": 0}
+                    global_key["metricsConditionalTagging"] = []
```
looks silly atm, this will be more meaningful when we make a proper default for globalconfig

TBS1996 committed:
wip

iker-barriocanal committed:
reformat code inside macro

iker-barriocanal committed:
change names

iker-barriocanal committed:
simplify authentication and error handling, and add some docs

iker-barriocanal committed:
make fetch interval configurable

iker-barriocanal committed:
remove `enabled` from global config service

iker-barriocanal committed:
rename globalconfiguration -> global config

Since there's a collision with the config object itself. Not convinced
on the naming though.

iker-barriocanal committed:
ignore inner channel errors

iker-barriocanal committed:
GlobalConfigMessage -> GlobalConfigInterface

Still not convinced on the name :(

iker-barriocanal committed:
Get rid of the internal channel in favor of passing the watch channel

iker-barriocanal committed:
GlobalConfigurationService -> GlobalConfigService

iker-barriocanal committed:
fix some docstrings

TBS1996 committed:
fix doc link

iker-barriocanal committed:
update some docs

iker-barriocanal committed:
stop processor service if no matches

olksdr commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,153 @@
+    // from the upstream are resolved faster than the fetch interval of this
+    // service. An alternative is to create a channel internal to the service
+    // and handle the updates through them.
+    sender: Arc<watch::Sender<Arc<GlobalConfig>>>,
```
I wouldn't use the `Arc` over `watch` here, and just either use another inner channel to send the result of the update to the inner loop and use the sender there, or use the `broadcast` channel from tokio

jan-auer replied:
That's overcomplicating it, I'm afraid. The sender offers an immutable interface exactly for use cases like this one where it needs to be updated from spawned tasks. Creating yet another channel and complicated loop is too much indirection for what this service actually does.

However, we might need a back channel due to https://github.com/getsentry/relay/pull/2320#discussion_r1289828027, in which case the `Arc` should be removed, too.

olksdr commented on relay-server/src/actors/processor.rs:
```diff
@@ -2750,22 +2764,38 @@ impl EnvelopeProcessorService {
+                .global_config
+                .send(Subscribe)
+                .await
+                .expect("EnvelopeProcessorService failed subscribing to GlobalConfigService");
```
That might be not a good idea to unwrap here

TBS1996 replied:
Do you suggest a way to keep going with the select! loop without the global config receiver, or just another way to shut down?

jan-auer replied:
Either that, or you even subscribe outside (in ServiceState::create), then clone the receiver, and just pass that in.

olksdr commented on relay-server/src/actors/processor.rs:
```diff
@@ -2750,22 +2764,38 @@ impl EnvelopeProcessorService {
+                    (Some(message), Ok(permit)) = async {tokio::join!(
+                        rx.recv(),
+                        semaphore.clone().acquire_owned()
+                    )} => {
```
```suggestion
                    (Some(message), Ok(permit)) =  futures::future::join(
                        rx.recv(),
                        semaphore.clone().acquire_owned()
                    ) => {

```

TBS1996 replied:
neat, didnt know about this one!

jan-auer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,153 @@
+    /// Upstream service to request global configs from.
+    upstream: Addr<UpstreamRelay>,
+    /// Number of seconds to wait before making another request.
+    fetch_interval: u64,
```
Please make this an `std::time::Duration`. A field `fetch_interval` does not convey which unit is in. Notice how we use `Duration` in all our interfaces (except for raw config structs, since it's not serializable; `Config` has accessor methods for this and clear documentation on all fields).

jan-auer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,153 @@
+}
+
+/// A way to get updates of the global config.
+pub enum GlobalConfigInterface {
```
Should this be called `GlobalConfigManager` instead? We don't use "Interface" as per naming convention.

jan-auer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,153 @@
+/// The service offers two alternatives to fetch the [`GlobalConfig`]:
+/// responding to a [`GlobalConfigInterface::Get`] message with the config for
+/// one-off requests, or subscribing to updates with
+/// [`GlobalConfigInterface::Subscribe`] to keep up-to-date.
```
Better link to the `Subscribe` type here, since that is what callers will use. They invoke this as `global_config.send(Subscribe)`.

jan-auer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,153 @@
+    }
+}
+
+impl GlobalConfigService {
```
Move the definition of `struct GlobalConfigService` immediately here above the impl block. As per [coding guidelines](https://develop.sentry.dev/rust/#declaration-order):

> When declaring a structure with an implementation, always ensure that the struct and its impl blocks are consecutive.

jan-auer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,153 @@
+impl GlobalConfigService {
+    /// Creates a new [`GlobalConfigService`].
+    pub fn new(upstream: Addr<UpstreamRelay>) -> Self {
+        let (sender, _) = watch::channel(Arc::new(GlobalConfig::default()));
```
```suggestion
        let (sender, _) = watch::channel(Arc::default());
```

jan-auer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,153 @@
+        Self {
+            sender: Arc::new(sender),
+            upstream,
+            fetch_interval: 10,
```
Please make this a field in Relay's config where other intervals are configured for the cache.

jan-auer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,153 @@
+            match upstream_relay.send(SendQuery(query)).await {
+                Ok(res) => Self::handle_upstream_response(res, sender),
+                Err(_) => relay_log::error!(
+                    "GlobalConfigService failed to send request to UpstreamService"
```
_nit_: You can shorten the error message here as the logger will automatically add the module and location.

jan-auer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,153 @@
+        response: Result<GetProjectStatesResponse, UpstreamRequestError>,
+        sender: Arc<watch::Sender<Arc<GlobalConfig>>>,
+    ) {
+        if let Ok(config) = response {
```
Shouldn't we log the error here (`error!(error = error as &dyn std::error::Error, "..."`) so we know what happened?

iker-barriocanal committed:
Introduce a state to not process envelopes without global config

Didn't update the tests and it currently doesn't compile.

jan-auer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,153 @@
+        tokio::spawn(async move {
+            relay_log::info!("global config service starting");
+
+            let mut ticker = tokio::time::interval(Duration::from_secs(self.fetch_interval));
```
This is racy, since the interval does not wait for the upstream request to succeed. If the upstream request takes longer than the interval, multiple concurrent requests will be scheduled. 

Please look at services like the `OutcomeAggregator` or `UpstreamProjectSource`. They use a `SleepHandle` to solve this situation. Here's an adapted implementation to get you started:

 1. Initialize an idle handle in the constructor
 2. Fetch immediately when spawning the service
 3. After fetching, call `schedule_fetch`

```rust
struct GlobalConfigService {
    // ...
    fetch_interval: Duration,
    fetch_handle: SleepHandle, // Initialize with ::idle() in constructor
}

impl GlobalConfigService {
    /// Schedules the next trigger for fetching the global config.
    fn schedule_fetch(&mut self) {
        if self.fetch_handle.is_idle() {
            self.fetch_handle.set(self.fetch_interval);
        }
    }
}

impl Service for GlobalConfigService {
    fn spawn_handler(mut self, mut rx: relay_system::Receiver<Self::Interface>) {
        tokio::spawn(async move {
            self.do_fetch();

            loop {
                tokio::select! {
                    biased;
                    () = &mut self.fetch_handle => self.do_fetch(),
                    // ...
                }
            }
        });
    }
}
```

jan-auer commented on relay-server/src/actors/processor.rs:
```diff
@@ -563,12 +574,14 @@ impl EnvelopeProcessorService {
         };
 
         Self {
+            global_config: Arc::new(GlobalConfig::default()),
```
```suggestion
            global_config: Arc::default(),
```

iker-barriocanal committed:
replace async block with join function

jan-auer commented on relay-server/src/actors/processor.rs:
```diff
@@ -522,13 +529,15 @@ impl FromMessage<RateLimitFlushBuckets> for EnvelopeProcessor {
     config: Arc<Config>,
     envelope_manager: Addr<EnvelopeManager>,
     project_cache: Addr<ProjectCache>,
+    global_config: Addr<GlobalConfigInterface>,
```
Does this really need to reside in the inner processor? This should be outside, as it's not needed to process envelopes.

TBS1996 replied:
I thought we wanted to put everything that we don't change into the inner, so that we only need to clone two Arcs, and we don't need to change the global_config addr

jan-auer commented on relay-server/src/actors/processor.rs:
```diff
@@ -2750,22 +2764,38 @@ impl EnvelopeProcessorService {
-                tokio::join!(rx.recv(), semaphore.clone().acquire_owned())
-            {
-                let service = self.clone();
+            let mut global_config_rx = self
```
In addition to retrieving the rx, you should await it once to ensure you have an up-to-date config. That is, unless you want to process with default config, but that can be dangerous. 

Of course, the GlobalConfigService should still be able to send you a default config under certain conditions (e.g. in proxy mode), but that's not a concern of the EnvelopeProcessor.

jan-auer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,153 @@
+use relay_system::{Addr, AsyncResponse, FromMessage, Interface, Sender, Service};
+use tokio::sync::watch;
+
+use crate::actors::project_upstream::{GetProjectStates, GetProjectStatesResponse};
```
It's not clean that we're using `project_upstream` types here, since we don't actually use that service. 
- Either create a minimal copy of the types that have just the fields we need
- Or move the types to another module

Especially that we're starting to use `pub(crate)` shows that we now get into strange API visibility. I'd heavily lean towards the first option and undo all changes to the upstream service file.

jan-auer commented on relay-server/src/actors/project_upstream.rs:
```diff
@@ -58,7 +64,7 @@ impl UpstreamQuery for GetProjectStates {
 
     fn path(&self) -> Cow<'static, str> {
-        Cow::Borrowed("/api/0/relays/projectconfigs/?version=3")
+        Cow::Borrowed("/api/0/relays/projectconfigs/?version=4")
```
**IMPORTANT**: Please leave this unchanged for now. Continue to request v3 for project configs and use v4 ONLY for global configs.

iker-barriocanal committed:
GlobalConfigInterface -> GlobalConfigManager

iker-barriocanal committed:
Link to message types in docs to match external POV

iker-barriocanal committed:
Place struct definition and implementation next to each other

TBS1996 committed:
fix minisentry

iker-barriocanal committed:
undo state

iker-barriocanal committed:
Require global config and subscription before starting to process envelopes

iker-barriocanal committed:
simplify default global config initialization

iker-barriocanal committed:
subscribe before get

iker-barriocanal committed:
reorder impl blocks

TBS1996 committed:
wip

TBS1996 committed:
wip

iker-barriocanal committed:
accept config over specific value

iker-barriocanal committed:
emit global config fetching from upstream errors

iker-barriocanal committed:
don't depend on projectupstream in globalconfigs

iker-barriocanal committed:
limit global config requests

iker-barriocanal committed:
add some docs

iker-barriocanal committed:
forward upstream response internally

also: fix some docs

iker-barriocanal committed:
undo project upstream changes

iker-barriocanal commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,225 @@
+                    }
+                }
+                // FIXME(iker): if the request to upstream fails, we should
+                // reschedule another fetch anyway.
```
Other Relay services have `SleepHandle`s and `RetryBackoff`s. These services use the backoff mechanism to know if a request was started, but they also assume a frequent number of incoming messages and use that chance to check the backoff. The global config service can't rely on that assumption due to the lower number of incoming messages, so I see two options here.

1. Accept more than one ongoing request at a time (not amazing).
2. Accept that no more global config updates will be sent if an upstream request fails (pretty bad).

Would love to see if somebody else has a different opinion or suggestions.

TBS1996 replied:
I think I solved it by passing a double result to `handle_upstream_response`. now it will also handle the case for when the sending itself fails, so that new global configs will still be fetched.  
 
Now it will only fail if the internal receiver is dropped before the sender, which should be impossible since they both have the same lifetime.

TBS1996 committed:
add publickeys to globalconfig query

TBS1996 committed:
fix merge

TBS1996 committed:
check for credentials before requesting global config

TBS1996 committed:
Continue fetching global configs if sending to upstream fails

TBS1996 commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,230 @@
+    /// Sender of the internal channel to forward global configs from upstream.
+    internal_tx: UnboundedSender<UpstreamResult>,
+    /// Receiver of the internal channel to forward global configs from upstream.
+    internal_rx: UnboundedReceiver<UpstreamResult>,
```
wouldn't this struct be a bit cleaner if we created the channel in `spawn_handler` and cloned the sender in the function call?

TBS1996 commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,230 @@
+    /// Schedules the next global config request.
+    fn schedule_fetch(&mut self) {
+        if self.fetch_handle.is_idle() {
+            // XXX(iker): set a backoff mechanism?
```
i don't really see why we would use a backoff mechanism, isn't that for cases where you'd otherwise have too many frequent retries? if it's just once every 10 seconds, that seems totally fine in my mind

TBS1996 commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,230 @@
+            let query = GetGlobalConfig::query();
+            let res = upstream_relay.send(SendQuery(query)).await;
+            if internal_tx.send(res).is_err() {
+                relay_log::error!("failed to forward internally the upstream response");
```
if this error happens, it would stop requesting global configs altogether, however, i think it's impossible for this to happen as according to the channel docs, it should only fail if the receiver is dropped, which it won't.

TBS1996 commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,230 @@
+            // NOTE(iker): if this first request fails it's possible the default
+            // global config is forwarded. This is not ideal, but we accept it
+            // for now.
+            self.update_global_config();
```
if we initiated the sleep_handle with 0 sec duration, then i think this line wouldn't be necessary, as it would request within the loop immediatly, what do you think?

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 committed:
wip

TBS1996 commented on relay-server/src/actors/processor.rs:
```diff
@@ -2733,27 +2743,63 @@ impl EnvelopeProcessorService {
+                std::process::exit(1);
+            }
+        }
+    }
```
@iker-barriocanal I refactored this piece of code you wrote, using join! now since the two requests shouldn't depend on each other.  
 
it would be nice to get the global config in the constructor, but I suppose that's not done because it would be an async operation, is that right?

TBS1996 committed:
Merge branch 'master' into tor/globalconfig-service

TBS1996 committed:
wip

iker-barriocanal committed:
minor global config service changes

iker-barriocanal committed:
Use bounded channel of size 1 & other minor changes

iker-barriocanal committed:
Don't exit the process when subscribing errors in processor

iker-barriocanal committed:
remove unused vars

iker-barriocanal committed:
minor cleanup

iker-barriocanal committed:
Merge branch 'master' into tor/globalconfig-service

iker-barriocanal committed:
empty global struct from master

iker-barriocanal committed:
update tests

jjbayer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,242 @@
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct GetGlobalConfig {
+    pub global: bool,
```
Does this need to be public?

```suggestion
struct GetGlobalConfig {
    global: bool,
```

jan-auer replied:
+1, better not to have structs with mixed visibility. Either the entire fields are public (a data struct), or there are accessor methods.

jjbayer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,242 @@
+}
+
+impl GetGlobalConfig {
+    pub fn query() -> GetGlobalConfig {
```
```suggestion
    pub fn new() -> GetGlobalConfig {
```

jjbayer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,242 @@
+pub struct GlobalConfigService {
+    config: Arc<Config>,
+    /// Sender of the [`watch`] channel for the subscribers of the service.
+    sender: watch::Sender<Arc<GlobalConfig>>,
```
Just a suggestion to make `handle_message` more readable:

```suggestion
    global_config: watch::Sender<Arc<GlobalConfig>>,
```

iker-barriocanal replied:
I'm not sure if `global_config` better represents the purpose of this variable. I think it could be confusing with a global config instance when reading the code below. I'm open to changing it but I can't think of better suggestions, maybe "subscriber"?

jjbayer replied:
Best to pick a name that makes sense from the point of view of the code that uses this field:

```rust
sender.send(self.sender.borrow().clone());
```

Maybe `global_config_watch`?

iker-barriocanal replied:
Sounds great! Renamed in [255f885](https://github.com/getsentry/relay/pull/2320/commits/255f885d8d7cca30f028dca14dd92cbe4e00d29b).

jjbayer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,242 @@
+            ),
+        }
+
+        self.schedule_fetch();
```
Is there a chance that through some error, `handle_result` does not get called (e.g. the upstream does not respond), meaning we never try to fetch config again? The previous implementation with a fetch attempt running every 10 seconds unconditionally seemed more robust to me.

iker-barriocanal replied:
I had the same thought. The only situation where `handle_result` is not called is when the internal channel is closed (see comment on line 174).

jjbayer commented on relay-server/src/endpoints/project_configs.rs:
```diff
@@ -84,12 +95,14 @@ struct GetProjectStatesResponseWrapper {
 #[derive(Debug, Deserialize)]
 #[serde(rename_all = "camelCase")]
-struct GetProjectStatesRequest {
+pub struct GetProjectStatesRequest {
```
Does this need to be `pub`?

jjbayer commented on tests/integration/test_projectconfigs.py:
```diff
@@ -211,9 +211,9 @@ def request_config():
+def request_config(relay, packed, signature, version: str):
     return relay.post(
-        "/api/0/relays/projectconfigs/?version=3",
+        "/api/0/relays/projectconfigs/?version=" + version,
```
nit

```suggestion
        f"/api/0/relays/projectconfigs/?version={version}",
```

jan-auer commented on relay-config/src/config.rs:
```diff
@@ -833,6 +833,8 @@ struct Cache {
     file_interval: u32,
     /// Interval for evicting outdated project configs from memory.
     eviction_interval: u32,
+    /// Interval for fetching new global configs from the upstream.
```
_nit_: Specify in the doc comment that this interval is in seconds.

jan-auer commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,242 @@
+    fn update_global_config(&mut self) {
+        self.fetch_handle.reset();
+
+        let upstream_relay: Addr<UpstreamRelay> = self.upstream.clone();
```
```suggestion
        let upstream_relay = self.upstream.clone();
```

jan-auer commented on tests/integration/fixtures/mini_sentry.py:
```diff
@@ -308,16 +308,24 @@ def get_project_config():
-        elif version in ["2", "3"]:
+        elif version in ["2", "3", "4"]:
+            if version == "4":
+                if flask_request.json["global"]:
```
Even in v4, this will be an optional field. Note how the other fields are required, hence we should still fail if they are missing.

```suggestion
                if flask_request.json.get("global"):
```

olksdr commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,242 @@
+use crate::utils::SleepHandle;
+
+/// The result of sending a global config query to upstream. It can fail both in sending it,
+/// and in the response.
```
```suggestion
/// The result of sending a global config query to upstream. 
///
/// It can fail both in sending it, and in the response.
```

olksdr commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,242 @@
+/// The message for receiving a watch that subscribes to the [`GlobalConfigService`].
+pub struct Subscribe;
+
+/// A way to get updates of the global config.
```
It would be great to add more docs about `GlobalConfigManager` here, describe what it does and why and what this interface represents.

olksdr commented on relay-server/src/actors/global_config.rs:
Top level docs would be great as well. At least a sentence or two. 
It would be also easier later to refactor the entire module into its own crate and all the docs will be there

olksdr commented on relay-server/src/actors/global_config.rs:
```diff
@@ -0,0 +1,242 @@
+            relay_log::info!("global config service stopped");
+        });
+    }
+}
```
It would be also great to add unit tests to the this service in the followups.

iker-barriocanal committed:
address some feedback

iker-barriocanal committed:
add docs

iker-barriocanal committed:
rename sender -> global_config_watch

TBS1996 committed:
Merge branch 'master' into tor/globalconfig-service

iker-barriocanal committed:
deal with the global config independently in minisentry
